src/XMakeTasks/AssemblyDependency/AssemblyFoldersFromConfig/AssemblyFoldersFromConfigResolver.cs

```diff
@@ -0,0 +1,297 @@
+
+                                // If the processor architecture does not match the we should continue to see if there is a better match.
+                                if (foundAssembly != null && (foundAssembly.AssemblyName.ProcessorArchitecture == ProcessorArchitecture.MSIL || foundAssembly.AssemblyName.ProcessorArchitecture == ProcessorArchitecture.None))
+                                {
+                                    foundPath = candidatePath;
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+
+                    // If we get to this point and have not returned then we have the best assembly we could find, lets return it.
+                    if (resolvedPath != null)
+                    {
+                        foundPath = resolvedPath;
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+    }
+
+    /// <summary>
+    /// Contains information about entries in the AssemblyFoldersEx registry keys.
+    /// </summary>
+    internal class AssemblyFoldersFromConfigCache
+    {
+        /// <summary>
+        /// Set of files in ALL AssemblyFolderFromConfig directories
+        /// </summary>
+        private readonly HashSet<string> _filesInDirectories = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// File exists delegate we are replacing
+        /// </summary>
+        private readonly FileExists _fileExists;
+
+        /// <summary>
+        /// Should we use the original on or use our own
+        /// </summary>
+        private readonly bool _useOriginalFileExists;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        internal AssemblyFoldersFromConfigCache(AssemblyFoldersFromConfig assemblyFoldersFromConfig, FileExists fileExists)
+        {
+            AssemblyFoldersFromConfig = assemblyFoldersFromConfig;
+            _fileExists = fileExists;
+
+            if (Environment.GetEnvironmentVariable("MSBUILDDISABLEASSEMBLYFOLDERSEXCACHE") != null)
+            {
+                _useOriginalFileExists = true;
+            }
+            else
+            {
+                var lockobject = new object();
+
+                Parallel.ForEach(assemblyFoldersFromConfig, assemblyFolder =>
+                {
+                    if (!FileUtilities.DirectoryExistsNoThrow(assemblyFolder.DirectoryPath)) return;
+
+                    string[] files = Directory.GetFiles(assemblyFolder.DirectoryPath, "*.*", SearchOption.TopDirectoryOnly);
+
+                    lock (lockobject)
+                    {
+                        foreach (var file in files)
+                        {
+                            _filesInDirectories.Add(file);
+                        }
```

Comment:
Actually this whole mess could be dramatically simplified:

```C#
_filesInDirectories = assemblyFoldersEx.AsParallel()
    .Where(assemblyFolder => FileUtilities.DirectoryExistsNoThrow(assemblyFolder.DirectoryPath))
    .SelectMany(
        assemblyFolder =>
            Directory.GetFiles(assemblyFolder.DirectoryPath, "*.*", SearchOption.TopDirectoryOnly))
    .ToImmutableHashSet(StringComparer.OrdinalIgnoreCase);
```

---
In reply to: [73542581](https://github.com/Microsoft/msbuild/pull/863#discussion_r73542581) [](ancestors = 73542581)